File: leetcode/graph/1615. Maximal Network Rank/js/main.js
// Time O(N^2)
// Space O(N)
const maximalNetworkRank = (n, roads) => {
  let adjList = new Map();
  let visited = new Set();

  for (let i = 0; i < n; i++) {
    adjList.set(i, new Set());
  }

  for (let [u, v] of roads) {
    adjList.get(u).add(v);
    adjList.get(v).add(u);
  }

  let max = 0;

  for (let i = 0; i < n; i++) {
    for (let j = i + 1; j < n; j++) {
      let city1 = adjList.get(i);
      let city2 = adjList.get(j);
      let diff = city2.has(i) ? 1 : 0;
      max = Math.max(max, city1.size + city2.size - diff);
    }
  }

  return max;
};

File: leetcode/graph/1615. Maximal Network Rank/py/main.py
import collections


class Solution:
    def maximalNetworkRank(self, n: int, roads: List[List[int]]) -> int:
        connected = collections.defaultdict(set)

        for a, b in roads:
            connected[a].add(b)
            connected[b].add(a)

        res = 0

        for i in range(n):
            for j in range(i + 1, n):
                city1 = connected[i]
                city2 = connected[j]
                diff = 1 if i in city2 else 0
                res = max(res, len(city1) + len(city2) - diff)

        return res
